neo/Ledger/MemoryPool.cs

```diff
@@ -1,165 +1,513 @@
 ﻿using Neo.Network.P2P.Payloads;
 using System;
 using System.Collections;
-using System.Collections.Concurrent;
 using System.Collections.Generic;
 using System.Linq;
-using System.Numerics;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using Akka.Util.Internal;
+using Neo.Network.P2P;
+using Neo.Persistence;
+using Neo.Plugins;
 
 namespace Neo.Ledger
 {
-    internal class MemoryPool : IReadOnlyCollection<Transaction>
+    public class MemoryPool : IReadOnlyCollection<Transaction>
     {
-        private class PoolItem
+        private class PoolItem : IComparable<PoolItem>
         {
             public readonly Transaction Transaction;
             public readonly DateTime Timestamp;
+            public DateTime LastBroadcastTimestamp;
 
             public PoolItem(Transaction tx)
             {
                 Transaction = tx;
                 Timestamp = DateTime.UtcNow;
+                LastBroadcastTimestamp = Timestamp;
+            }
+
+            public int CompareTo(Transaction tx)
+            {
+                if (tx == null) return 1;
+                int ret = Transaction.FeePerByte.CompareTo(tx.FeePerByte);
+                if (ret != 0) return ret;
+                ret = Transaction.NetworkFee.CompareTo(tx.NetworkFee);
+                if (ret != 0) return ret;
+
+                return Transaction.Hash.CompareTo(tx.Hash);
+            }
+
+            public int CompareTo(PoolItem otherItem)
+            {
+                if (otherItem == null) return 1;
+                return CompareTo(otherItem.Transaction);
             }
         }
 
-        private readonly ConcurrentDictionary<UInt256, PoolItem> _mem_pool_fee = new ConcurrentDictionary<UInt256, PoolItem>();
-        private readonly ConcurrentDictionary<UInt256, PoolItem> _mem_pool_free = new ConcurrentDictionary<UInt256, PoolItem>();
+        // Allow reverified transactions to be rebroadcast if it has been this many block times since last broadcast.
```

Comment:
I’m going to switch this to use a linear formula that increases if their are more transactions in the pool since rebroadcasting has more network load when the pool is more full.

Reply:
I’ll do it in a separate PR though. I think it is important to get this change merged in now.